### ISSUE TITLE
Gemless

### ISSUE THREAD

**Issue by sj26** [sj26/mailcatcher#298]
Gems included with MailCatcher, especially ones with native extensions, are the core of many users problems. I wonder if we can create an entirely gemless version, or at least something which vendors everything.

**Comment by sj26** [comment]
This is more and more necessary given how EventMachine, Thin, and the whole ruby world have moved on.

Refactoring into a pure-ruby threaded smtp + http server with some basic websocket support would probably be fine. I might take a look at this tonight while I'm on a plane. I've been procrastinating for far too long.

**Comment by luckman212** [comment]
I guess I'm 3 years late to the party but I just discovered mailcatcher and think it's great. I was getting the `Fixnum is deprecated` error when starting, so I started digging and tried updating `thin` (to 1.7.2). I saw that `skinny` is only tested up to thin 1.6 but I manually edited the dependencies and mailcatcher seems to be doing fine so far, despite #226 and https://github.com/sj26/skinny/pull/3.

Here's my sledgehammer method (tested on macOS 10.14.6 + ruby 2.6.3p62 via [rbenv](https://github.com/rbenv/rbenv))
```
$ gem install mailcatcher
$ gem install thin -v 1.7.2
$ gem uninstall -I thin -v 1.5.1
$ gem cleanup

$ sed -i .bak 's/^gem "thin", "~> 1.5.0"$/gem "thin", "~> 1.7"/' ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/mailcatcher-0.7.1/lib/mail_catcher.rb
$ sed -i .bak 's/_dependency(%q<thin>.freeze, \["~> 1.5.0"\]/_dependency(%q<thin>.freeze, \["~> 1.7"\]/' ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/specifications/mailcatcher-0.7.1.gemspec
$ sed -i .bak 's/_dependency(%q<thin>.freeze, \[">= 1.5", "< 1.7"\]/_dependency(%q<thin>.freeze, [">= 1.5"]/' ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/specifications/skinny-0.2.4.gemspec
```
Am I nuts for doing this?

**Comment by zarelit** [comment]
@sj26 I think that having mailcatcher as a gem is great because you can just `gem install` and have the ruby ecosystem carry the setup out for you, the main downside is that you *can* add it to the Gemfile along with all of its dependencies constraints.
IMHO mailcatcher is to the end user more of an application (or a service) rather than a library and thus should be either installed via the distribution package manager or through a container system like docker and let the users know that they can install it through gem if they know how it works.

**Comment by sj26** [comment]
I wholeheartedly agree. But there are so many issues caused by folks bundle installing it that I’d love to make it just work without gems. That said, it’s probably not hard, and not worthwhile.

**Comment by zarelit** [comment]
@sj26 well at least now github gives us a metric of how many people are not reading the documentation:

**Comment by zykadelic** [comment]
Trying to install this on a new system (macOS 10.15.6) currently fails for me. On running `rvm default@mailcatcher --create do gem install mailcatcher`, I get this output (same result with `install --pre`):
```
Fetching rack-protection-1.5.5.gem
Fetching daemons-1.3.1.gem
Fetching rack-1.6.13.gem
Fetching eventmachine-1.0.9.1.gem
Fetching thin-1.5.1.gem
Fetching skinny-0.2.4.gem
Fetching tilt-2.0.10.gem
Fetching sqlite3-1.4.2.gem
Fetching sinatra-1.4.8.gem
Fetching mini_mime-1.0.2.gem
Fetching mail-2.7.1.gem
Fetching mailcatcher-0.7.1.gem
Successfully installed daemons-1.3.1
Building native extensions. This could take a while...
Successfully installed eventmachine-1.0.9.1
Successfully installed rack-1.6.13
Building native extensions. This could take a while...
ERROR:  Error installing mailcatcher:
	ERROR: Failed to build gem native extension.

    current directory: /Users/zykadelic/.rvm/gems/ruby-2.7.0@mailcatcher/gems/thin-1.5.1/ext/thin_parser
/Users/zykadelic/.rvm/rubies/ruby-2.7.0/bin/ruby -I /Users/zykadelic/.rvm/rubies/ruby-2.7.0/lib/ruby/2.7.0 -r ./siteconf20200921-31902-brn25x.rb extconf.rb
checking for main() in -lc... yes
creating Makefile

current directory: /Users/zykadelic/.rvm/gems/ruby-2.7.0@mailcatcher/gems/thin-1.5.1/ext/thin_parser
make "DESTDIR=" clean

current directory: /Users/zykadelic/.rvm/gems/ruby-2.7.0@mailcatcher/gems/thin-1.5.1/ext/thin_parser
make "DESTDIR="
compiling parser.c
parser.rl:112:17: warning: comparison of integers of different signs: 'long' and 'unsigned long' [-Wsign-compare]
  assert(pe - p == len - off && "pointers aren't same distance");
         ~~~~~~ ^  ~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
parser.rl:142:7: error: implicit declaration of function 'thin_http_parser_has_error' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  if (thin_http_parser_has_error(parser) ) {
      ^
parser.rl:142:7: note: did you mean 'http_parser_has_error'?
./parser.h:44:5: note: 'http_parser_has_error' declared here
int http_parser_has_error(http_parser *parser);
    ^
parser.rl:144:14: error: implicit declaration of function 'thin_http_parser_is_finished' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  } else if (thin_http_parser_is_finished(parser) ) {
             ^
1 warning and 2 errors generated.
make: *** [parser.o] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/zykadelic/.rvm/gems/ruby-2.7.0@mailcatcher/gems/thin-1.5.1 for inspection.
Results logged to /Users/zykadelic/.rvm/gems/ruby-2.7.0@mailcatcher/extensions/x86_64-darwin-19/2.7.0/thin-1.5.1/gem_make.out
```
Mailcatcher has saved me so much time while developing emails in the past and I've loved using it!

**Comment by sentience** [comment]
@zykadelic This issue can be worked around by adding a compiler flag:

```
gem install mailcatcher  -- --with-cflags="-Wno-error=implicit-function-declaration"
```

To get this working with bundler, you will need to add the same flag as a [build option](https://bundler.io/v1.17/bundle_config.html#BUILD-OPTIONS) to your bundler config.

**Comment by sj26** [comment]
Closed in favour of #467.
